crates/core/src/event.rs

```diff
@@ -13,21 +14,86 @@ use thiserror::Error;
 
 use crate::borsh::{BorshDeserialize, BorshSerialize};
 use crate::collections::HashMap;
-use crate::ethereum_structs::{BpTransferStatus, EthBridgeEvent};
+use crate::ethereum_structs::EthBridgeEvent;
 use crate::ibc::IbcEvent;
 
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __event_type_impl {
+    ($domain:ty) => {
+        <$domain as $crate::event::EventToEmit>::DOMAIN
+    };
+    ($domain:ty, $($subdomain:expr),*) => {
+        ::konst::string::str_join!(
+            "/",
+            &[
+                $crate::__event_type_impl!($domain),
+                $($subdomain),*
+            ],
+        )
+    };
+    // TODO: remove these variants of the macro
+    ($domain:expr) => {
+        $domain
+    };
+    ($domain:expr, $($subdomain:expr),*) => {
+        ::konst::string::str_join!(
+            "/",
+            &[
+                $crate::__event_type_impl!($domain),
+                $($subdomain),*
+            ],
+        )
+    };
+}
+
+/// Instantiate a new [`EventType`] in const contexts. Mostly
+/// useful to define new event types in the protocol.
+///
+/// # Example
+///
+/// ```ignore
+/// const RELAYED: EventType = event_type!(EthBridgeEvent, "bridge-pool", "relayed");
+/// ```
+#[macro_export]
+macro_rules! event_type {
+    ($($tt:tt)*) => {
+        $crate::event::EventType::new($crate::__event_type_impl!($($tt)*))
+    };
+}
+
+/// An event to be emitted in Namada.
+pub trait EventToEmit: Into<Event> {
+    /// The domain of the event to emit.
+    ///
+    /// This may be used to group events of a certain kind.
+    const DOMAIN: &'static str;
+}
+
+impl EventToEmit for Event {
+    const DOMAIN: &'static str = "unknown";
+}
+
+impl EventToEmit for IbcEvent {
+    const DOMAIN: &'static str = "ibc";
```

Comment:
The event check in Hermes assumes that all IBC events don't have a prefix.
The check is implemented independently from chain-specific implementation.
We could support these events in our fork Hermes, but it is difficult for the upstream.